### PR TITLE
GC: Fix CheckPromoted for frozen objects

### LINKED ISSUE #76219: Tests fail in x86 after the latest update with "Internal CLR error. (0x80131506)"
### Description

It looks like the latest automated flow https://github.com/dotnet/winforms/pull/7833 has brought a regression that causes our x86 steps to fail consistently.
For example:
* https://dev.azure.com/dnceng-public/public/_build/results?buildId=30791&view=results
* https://dev.azure.com/dnceng-public/public/_build/results?buildId=29338&view=results

### Reproduction Steps

Presumably it should be replicatable by running `.\build.cmd -test -platform x86` but, unfortuntately, I am unable to repro the crash locally. And I'm not really sure how to get any more information off the build agents.

### Expected behavior

Not crash like this, and if there's an issue with the tests the root cause of issue should clear.

### Actual behavior

x86 test runs fail
```
Fatal error. Internal CLR error. (0x80131506)
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, Void**, System.Signature, Boolean)
   at System.Reflection.ConstructorInvoker.Invoke(System.Object, IntPtr*, System.Reflection.BindingFlags)
   at System.Reflection.RuntimeConstructorInfo.Invoke(System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at System.Reflection.ConstructorInfo.Invoke(System.Object[])
   at Xunit.Sdk.XunitTestClassRunner+<>c__DisplayClass10_0.<CreateClassFixture>b__3()
   at Xunit.Sdk.ExceptionAggregator.Run(System.Action)
   at Xunit.Sdk.XunitTestClassRunner.CreateClassFixture(System.Type)
   at Xunit.Sdk.XunitTestClassRunner+<CreateClassFixtureAsync>d__11.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.XunitTestClassRunner+<CreateClassFixtureAsync>d__11, xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<CreateClassFixtureAsync>d__11 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.XunitTestClassRunner+<CreateClassFixtureAsync>d__11, xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<CreateClassFixtureAsync>d__11 ByRef)
   at Xunit.Sdk.XunitTestClassRunner.CreateClassFixtureAsync(System.Type)
   at Xunit.Sdk.XunitTestClassRunner+<AfterTestClassStartingAsync>d__13.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.XunitTestClassRunner+<AfterTestClassStartingAsync>d__13, xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<AfterTestClassStartingAsync>d__13 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.XunitTestClassRunner+<AfterTestClassStartingAsync>d__13, xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<AfterTestClassStartingAsync>d__13 ByRef)
   at Xunit.Sdk.XunitTestClassRunner.AfterTestClassStartingAsync()
   at Xunit.Sdk.TestClassRunner`1+<RunAsync>d__37[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestClassRunner`1+<RunAsync>d__37[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__37<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.TestClassRunner`1+<RunAsync>d__37[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__37<System.__Canon> ByRef)
   at Xunit.Sdk.TestClassRunner`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].RunAsync()
   at Xunit.Sdk.XunitTestCollectionRunner.RunTestClassAsync(Xunit.Abstractions.ITestClass, Xunit.Abstractions.IReflectionTypeInfo, System.Collections.Generic.IEnumerable`1<Xunit.Sdk.IXunitTestCase>)
   at Xunit.Sdk.TestCollectionRunner`1+<RunTestClassesAsync>d__28[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestCollectionRunner`1+<RunTestClassesAsync>d__28[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunTestClassesAsync>d__28<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.TestCollectionRunner`1+<RunTestClassesAsync>d__28[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunTestClassesAsync>d__28<System.__Canon> ByRef)
   at Xunit.Sdk.TestCollectionRunner`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].RunTestClassesAsync()
   at Xunit.Sdk.TestCollectionRunner`1+<RunAsync>d__27[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestCollectionRunner`1+<RunAsync>d__27[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__27<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.TestCollectionRunner`1+<RunAsync>d__27[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__27<System.__Canon> ByRef)
   at Xunit.Sdk.TestCollectionRunner`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].RunAsync()
   at Xunit.Sdk.XunitTestAssemblyRunner.RunTestCollectionAsync(Xunit.Sdk.IMessageBus, Xunit.Abstractions.ITestCollection, System.Collections.Generic.IEnumerable`1<Xunit.Sdk.IXunitTestCase>, System.Threading.CancellationTokenSource)
   at Xunit.Sdk.XunitTestAssemblyRunner+<>c__DisplayClass14_2.<RunTestCollectionsAsync>b__2()
   at System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
   at System.Threading.Tasks.Task+<>c.<.cctor>b__273_0(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
   at System.Threading.Tasks.Task.ExecuteEntry()
   at System.Threading.Tasks.SynchronizationContextTaskScheduler+<>c.<.cctor>b__8_0(System.Object)
   at Xunit.Sdk.MaxConcurrencySyncContext.RunOnSyncContext(System.Threading.SendOrPostCallback, System.Object)
   at Xunit.Sdk.MaxConcurrencySyncContext+<>c__DisplayClass11_0.<WorkerThreadProc>b__0(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at DynamicClass.lambda_method2(System.Runtime.CompilerServices.Closure, System.Object, System.Object)
   at Xunit.Sdk.MaxConcurrencySyncContext.WorkerThreadProc()
   at Xunit.Sdk.XunitWorkerThread+<>c.<QueueUserWorkItem>b__5_0(System.Object)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task+<>c.<.cctor>b__273_0(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
   at System.Threading.Tasks.ThreadPoolTaskScheduler+<>c.<.cctor>b__10_0(System.Object)
   at System.Threading.Thread.StartCallback()
```

### Regression?

It looks like a regression since nothing in our codebase changed.

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/76219

The problem is - when GC scans SyncBlocks in [GCWeakPtrScanElement](https://github.com/dotnet/runtime/blob/94c6fe6cf5667967f6cf0ebc5c381501b8363b9b/src/coreclr/vm/syncblk.cpp#L1226) and meets a SyncBlock holding a weak pointer to a frozen object (`RuntimeType` in my case) it invokes [this handler](https://github.com/dotnet/runtime/blob/94c6fe6cf5667967f6cf0ebc5c381501b8363b9b/src/coreclr/vm/syncblk.cpp#L1274) which is effectively [CheckPromoted](https://github.com/dotnet/runtime/blob/94c6fe6cf5667967f6cf0ebc5c381501b8363b9b/src/coreclr/gc/gcscan.cpp#L109-L126) where that weak ref is recognized as unreachable because it doesn't pass `if (!g_theGCHeap->IsPromoted(*pRef))` so the syncblock is cleared/removed/reused.

cc @jkotas @cshung @Maoni0 

Easily reproduces on this small snippet on Windows-x86 Checked - https://github.com/dotnet/runtime/issues/76219#issuecomment-1259455269
(not on the latest Main since frozen type objects were just reverted in https://github.com/dotnet/runtime/pull/76235 because of this issue)

### CODE REVIEW

**Comment by EgorBo** [comment]
Ah, I assume it either has to be moved to `gc.cpp` to `bool GCHeap::IsPromoted(Object* object)` or we can see if simply adding `| 1` bit to a ref in SyncBlock on creating is enough? I feel like the first option is safer?

**Comment by jkotas** [comment]
I am not sure about the best place to add this check. I would like to hear @Maoni0's opinion.

I suspect that we may have similar problem for weak handles and dependent handles that will also lead to bad failure modes.

**Comment by EgorBo** [comment]
Looks similar to my previous change in gc.cpp - https://github.com/dotnet/runtime/pull/76251/files

**Comment by EgorBo** [comment]
@jkotas we had a chat with @Maoni0 and I have an impression that the optimization that we decided to use - avoid managed pinned handles (for both RuntimeType and string literals) is the reason. Previously, all RuntimeType objects were handled by that pinned  table so RuntimeType objects were always marked while now e.g. RuntimeType can be completely unused for some period of time and then used again (with header containing out-of-date syncblock index)

**Comment by EgorBo** [comment]
Repro with strings:
```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using System;
using System.Runtime.CompilerServices;

public class Class1
{
    public static void Main()
    {
        Test();
        // "hello" is unused now, SyncBlock is cleared
        GC.Collect();

        // Now we need "hello" again!
        Test();
    }

    static void Test()
    {
        Say("hello");
    }
    static void Say(string str)
    {
        lock (str)
        {
            Console.WriteLine(str);
        }
    }
}
```

**Comment by EgorBo** [comment]
Possible solutions (after some chatting with Maoni):
1) Clear ObjectHeader once we collect SyncBlocks for frozen objects - it might work but it likely won't help with other potential problems like WeakReference to a RuntimeType object that goes unused for some period of time (and then used again)
2) Bring back pinned handles - it will also slightly simplify code, e.g. for string literals we had to do some additional changes to have unmanaged handles.
3) Same as 2) but having a pinned table (or registry or whatever name) in FOH itself and it will be rooted by `m_pPinnedHeapHandleTable` - more work than 2)

**Comment by EgorBo** [comment]
Implemented 3) option

**Comment by cshung** [comment]
I am not sure I understand, but are we saying we are going to mark/unmark all frozen objects for every GC as if they are roots?

**Comment by jkotas** [comment]
The fix for this problem should also cover objects from file-mapped frozen heaps. File-mapped frozen heaps do not have GC handles or an extra array pointing to the frozen heap objects.

It looks like the whole point of registry that you have implemented is to mark every single object on FOH. It is cheaper to just walk FOH and mark everything on it directly. We do not need an extra array to do that. What prevents us from doing that if we believe that marking all objects on FOH during the GC is the right tradeoff?

**Comment by EgorBo** [comment]
> I am not sure I understand, but are we saying we are going to mark/unmark all frozen objects for every GC as if they are roots?

Is there a better solution? We used to mark/unmark the same objects before FOH too so the main savings here are jit that is able to bake direct references to various objects in codegen + omit write barriers barriers when we assign them to fields/arrays. Also, we allocate less of pinned handles in VM.

My initial fix was `IsFrozen` check in gc.cpp's `bool GCHeap::IsPromoted(Object* object)` but after a chat with @Maoni0 I changed it to a safer option - root all FOH objects.

Basically, I want to fix at least two cases:
1) SyncBlocks for FOH objects should not be ever cleared (or FOH Objects' Headers have to be reset too)
2) WeakReference pointing to a FOH object should never report it as dead even if nobody uses given FOH object at a given time.
Same for dependent handlers I assume.

**Comment by EgorBo** [comment]
> It looks like the whole point of registry that you have implemented is to mark every single object on FOH. It is cheaper to just walk FOH and mark everything on it directly. We do not need an extra array to do that. What prevents us from doing that if we believe that marking all objects on FOH during the GC is the right tradeoff?

Yes, that's the point. What you suggest makes sense to me but I am not sure I understand how to do it, I assume somewhere in gc.cpp at mark phase GC should just walk all objects in FOH and mark, right?

**Comment by Maoni0** [comment]
no, we do not mark every object on FOH. they are only marked (if anyone refers to them) on the FOH segments that are *in range*. this naturally happens with GC's mark process.

**Comment by EgorBo** [comment]
> no, we do not mark every object on FOH. they are only marked (if anyone refers to them) on the FOH segments that are _in range_. this naturally happens with GC's mark process.

Right, but previously all the objects we moved now to FOH were always marked because they were referenced by some special array objects in VM. Now these objects are not handled by any special internal arrays/handlers, and instead are saved to some unmanaged data structures, e.g. string literal hash table or `size_t ExposedManagedObject` field in MethodTable so they're not marked any more in some cases while technically still alive and can be used again

**Comment by cshung** [comment]
Oh, this is a segment-only problem? Frozen segments are always out of range for regions?

**Comment by EgorBo** [comment]
> Oh, this is a segment-only problem? Frozen segments are always out of range for regions?

I might be wrong but the problem only reproduces with segments yes, the gcstress crash we had + winforms issue are both x86 where regions are not enabled

**Comment by cshung** [comment]
Just in case it is a segment-only issue, we can potentially optimize away all these in case we are running under regions, which is going to be the default anyway. Setting the mark bit or not, having the `mark_phase` to spend time proportional to the number of frozen objects (just because of sync block and weak references) is unfortunate.

**Comment by EgorBo** [comment]
> Just in case it is a segment-only issue, we can potentially optimize away all these in case we are running under regions, which is going to be the default anyway. Setting the mark bit or not, having the `mark_phase` to spend time proportional to the number of frozen objects (just because of sync block and weak references) is unfortunate.

Makes perfect sense to me, for Regions we don't drop SyncBlock because IsPromoted returns true since it's out of reach.
So are we fine with the current PR for segments only? I assume it doesn't add additional work for GC in comparison with e.g. .NET 7.0 since the number of objects to mark is the same. Or we still want that optimization Jan suggested for non-default case? In that case I'll probably need some guidance 🙂

**Comment by filipnavara** [comment]
> Ah, Jan meant that the current PR doesn't solve the same problem for e.g. NativeAOT so I guess it's not an option then.

FWIW NativeAOT currently doesn't have regions enabled on any platform (#69108).

**Comment by EgorBo** [comment]
> > Ah, Jan meant that the current PR doesn't solve the same problem for e.g. NativeAOT so I guess it's not an option then.
> 
> FWIW NativeAOT currently doesn't have regions enabled on any platform (#69108).

Right, so technically I can reproduce the same crash on .net 7.0 NativeAOT I believe. it's just that it's not easy without GCStress

**Comment by filipnavara** [comment]
> it's just that it's not easy without GCStress

Technically there's some GC Stress code in NativeAOT in Debug builds. I just never really tried to use it.

**Comment by jkotas** [comment]
> Setting the mark bit or not, having the mark_phase to spend time proportional to the number of frozen objects (just because of sync block and weak references) is unfortunate.

We do unmark all objects in frozen segments that are in-range in `sweep_ro_segments` at the end. Can we do similar marking for all objects in frozen segments that are in-range at the beginning?

**Comment by jkotas** [comment]
I assume that the crash happens only if we have in-range frozen segments. Out-of-range frozen segments are fine. Is this correct?

**Comment by jkotas** [comment]
> Right, so technically I can reproduce the same crash on .net 7.0 NativeAOT I believe.

NativeAOT does not use syncblocks. I do not think you would see the same crash. You may be able to observe weak handles or conditional handles to frozen objects getting cleared prematurely. It may cause subtle functional issues, but no crashes in the core runtime.

**Comment by EgorBo** [comment]
> I assume that the crash happens only if we have in-range frozen segments. Out-of-range frozen segments are fine. Is this correct?

At least that is my understanding. It used to return false here (because was in-range and unmarked): 

Screenshot 2022-09-28 at 05 25 13

**Comment by EgorBo** [comment]
So I mostly copy-pased `sweep_ro_segments` that @jkotas mentioned to implement "mark everything in in-range ro segments" but I am not sure I did it correctly, also I don't understand how to properly implement it for background gc (`seg_mark_array_bits_soh`) can you help me please?

**Comment by EgorBo** [comment]
@cshung @Maoni0 so when you have time could you please review my gc.cpp changes? thanks in advance

**Comment by Maoni0** [comment]
let's have a quick meeting to discuss this and get to a final solution instead of trying various things.

**Comment by cshung** [comment]
I think the original intent and promise for frozen segments is that these objects are *NOT* managed by the GC at all, we just did minimal work to make the runtime happy with that (e.g. the Write Barrier won't crash, we don't accidentally leave mark bits, ...). It would be unfortunate if we break that intent and spent time proportional to the number of frozen objects per GC.

In the abstract, the bug is that the GC is inadvertently freeing the "associated resources" of a frozen object if it is "in range". A natural way of fixing this should be to avoid that from happening.

> The "associated resource" in this particular case is the sync block, but it could also be weak references or dependent handle (or others).

The problem with fixing this directly is that it might be expensive to check if the "associated resource" belongs to a frozen object or not. This issue could be side-stepped if the "associated resource" were labeled as such.

With some thought, it is not hard to see the labeling must be done by the VM. (It would be meaningless if the GC needs to run the expensive check during GC to determine the label).

If we were changing the VM anyway, we also have the opportunity to seggregate those resources so that they are not presented to the GC at all. Suppose when the sync-block were created, we knew they belong to frozen objects, and so we store them somewhere else, then the GC wouldn't be looping through them, truly achieving frozen objects has zero GC cost.

> This is assuming we don't create those "associated resources" too often, which may or may not be true. 

Weak Reference and Dependent Handle are particularly easy to handle if we were to change the VM. We can simply replace that with a strong handle underneath when we detected it is a frozen object, and we are done. A dependent handle depending on a frozen object is just really dumb and costly.

**Comment by jkotas** [comment]
> It would be unfortunate if we break that intent and spent time proportional to the number of frozen objects per GC.

That intent is broken already (see the existing `seg_clear_mark_bits` method). You are spending time proportional to the number of (in-range) frozen objects per full GC already. Fortunately, this happens rarely with segments or never with regions. I do not think that it should be a goal for frozen objects to have zero GC cost. 

>  Suppose when the sync-block were created, we knew they belong to frozen objects, and so we store them somewhere else, then the GC wouldn't be looping through them, truly achieving frozen objects has zero GC cost.

It is still valuable to free the non-precious syncblocks, even when get attached to the frozen object. I do not see a problem here.

> A dependent handle depending on a frozen object is just really dumb and costly.

Well, we would need to introduce new types of handles to implement your idea. The new type of handle is somewhat expensive too. It would be worth of only if the dependent handles pointing to frozen objects are very common. I am pretty sure that it is not the case. I believe that having a few dependent handles pointing to frozen objects is the best among the available options.

**Comment by EgorBo** [comment]
So presumably the question is where to fix the problem: on GC side or VM (two VMs) side

> It is still valuable to free the non-precious syncblocks, even when get attached to the frozen object. I do not see a problem here.

Do I understand you correctly that you want to leave everything as is (as of Main) and fix the problem on VM side e.g.:
1) Clear object's header when we recycle a syncblock for FOH objects
2) Somehow make all WeakReferences to frozen objects strong ones
3) Same for dependent handles?
4) If we introduce a new entity in VM - will we catch potential bugs if we forget to handle it for FOH?
5) Do all of the above for NativeAOT (besides SyncBlocks which don't exist there)

At least this is how I understand the desire to collect SyncBlocks for frozen objects

**Comment by jkotas** [comment]
The change you have right should allow the non-precious syncblocks to be collected.

I like the change you have in the PR right now. I think it is the best path to fix this issue.

**Comment by jkotas** [comment]
> SyncBlocks should be rare beasts for frozen objects, right?

Right. SyncBlocks are rare in general.

The non-precious syncblocks handle the case where you have a little bit of contention on a long-lived instance, syncblock gets created and then never used again. We do not want to be stuck with this syncblock forever. We want to be able to free it while the long-lived instance is still alive.

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -34902,6 +34968,20 @@ void gc_heap::background_mark_phase ()
...
+    assert (!ro_segments_in_range);
+#else //USE_REGIONS
+    if (ro_segments_in_range)
+    {
+        dprintf (2, ("nonconcurrent marking in range ro segments"));
+        mark_ro_segments();
+        //concurrent_print_time_delta ("nonconcurrent marking in range ro segments");
+        concurrent_print_time_delta ("NRRO");
+    }
+#endif USE_REGIONS
```
Comment (jkotas) [inline src/coreclr/gc/gc.cpp:34982]:
```suggestion
#endif //USE_REGIONS
```

File: src/coreclr/gc/gc.cpp [hunk 2]
```diff
@@ -26388,6 +26396,19 @@ void gc_heap::mark_phase (int condemned_gen_number, BOOL mark_only_p)
...
+#ifdef USE_REGIONS
+        assert (!ro_segments_in_range);
+#else //USE_REGIONS
+        if (ro_segments_in_range)
+        {
+            dprintf(3,("Marking in range ro segments"));
+            mark_ro_segments();
+            // Should fire an ETW event here.
+        }
+#endif USE_REGIONS
```
Comment (jkotas) [inline src/coreclr/gc/gc.cpp:26409]:
```suggestion
#endif //USE_REGIONS
```

File: src/coreclr/gc/gc.cpp (hunk 1 above)
Comment (jkotas) [inline src/coreclr/gc/gc.cpp:34982]:
```suggestion
#endif //USE_REGIONS
```

**Comment by EgorBo** [comment]
Thanks for changes! Looks like there are two test failures I am able to reproduce locally - a valid frozen string literal is not marked when `IsPromoted` is called, here is the trace:

```
 	coreclr.dll!DbgAssertDialog(const char * szFile, int iLine, const char * szExpr) Line 461	C++
>	coreclr.dll!WKS::GCHeap::IsPromoted(Object * object) Line 45284	C++
 	coreclr.dll!CheckPromoted(Object * * pObjRef, unsigned int * pExtraInfo, unsigned int lp1, unsigned int lp2) Line 325	C++
 	coreclr.dll!ScanConsecutiveHandlesWithoutUserData(Object * * pValue, Object * * pLast, ScanCallbackInfo * pInfo, unsigned int * __formal) Line 446	C++
 	coreclr.dll!BlockScanBlocksWithoutUserData(TableSegment * pSegment, unsigned int uBlock, unsigned int uCount, ScanCallbackInfo * pInfo) Line 557	C++
 	coreclr.dll!SegmentScanByTypeMap(TableSegment * pSegment, const int * rgTypeInclusion, void(__stdcall*)(TableSegment *, unsigned int, unsigned int, ScanCallbackInfo *) pfnBlockHandler, ScanCallbackInfo * pInfo) Line 1666	C++
 	coreclr.dll!TableScanHandles(HandleTable * pTable, const unsigned int * puType, unsigned int uTypeCount, TableSegment *(__stdcall*)(HandleTable *, TableSegment *, Holder<CrstBase *,&CrstBase::AcquireLock,&CrstBase::ReleaseLock,0,&CompareDefault<CrstBase *>,1> *) pfnSegmentIterator, void(__stdcall*)(TableSegment *, unsigned int, unsigned int, ScanCallbackInfo *) pfnBlockHandler, ScanCallbackInfo * pInfo, Holder<CrstBase *,&CrstBase::AcquireLock,&CrstBase::ReleaseLock,0,&CompareDefault<CrstBase *>,1> * pCrstHolder) Line 1725	C++
 	coreclr.dll!HndScanHandlesForGC(HandleTable * hTable, void(__stdcall*)(Object * *, unsigned int *, unsigned int, unsigned int) scanProc, unsigned int param1, unsigned int param2, const unsigned int * types, unsigned int typeCount, unsigned int condemned, unsigned int maxgen, unsigned int flags) Line 790	C++
 	coreclr.dll!Ref_CheckAlive(unsigned int condemned, unsigned int maxgen, unsigned int lp1) Line 1408	C++
 	coreclr.dll!GCScan::GcShortWeakPtrScan(int condemned, int max_gen, ScanContext * sc) Line 143	C++
 	coreclr.dll!WKS::gc_heap::background_mark_phase() Line 35053	C++
 	coreclr.dll!WKS::gc_heap::gc1() Line 20981	C++
 	coreclr.dll!WKS::gc_heap::bgc_thread_function() Line 36016	C++
 	coreclr.dll!WKS::gc_heap::bgc_thread_stub(void * arg) Line 33961	C++
 	coreclr.dll!`anonymous-namespace'::CreateSuspendableThread::__l2::<lambda>(void * argument) Line 1415	C++
 	coreclr.dll!unsigned long <lambda>(void *)::<lambda_invoker_stdcall>(void * argument) Line 1419	C++
 	[External Code]	
 	[Frames below may be incorrect and/or missing, no symbols loaded for kernel32.dll]	
```

File: src/coreclr/gc/gc.cpp [hunk 3]
```diff
@@ -11009,7 +11009,7 @@ void gc_heap::seg_set_mark_array_bits_soh (heap_segment* seg)
     if (bgc_mark_array_range (seg, FALSE, &range_beg, &range_end))
     {
         size_t beg_word = mark_word_of (align_on_mark_word (range_beg));
-        size_t end_word = mark_word_of (align_on_mark_word (range_beg));
+        size_t end_word = mark_word_of (align_on_mark_word (range_end));
```
Comment (drieseng) [inline src/coreclr/gc/gc.cpp:11012]: You pasted that commit message from another PR, right? :-)

**Comment by EgorBo** [comment]
@Maoni0 @cshung the tests seem to be passing now (except `CoreCLR Product Build FreeBSD x64 release` that doesn't look to be related) so please approve if it looks good (modulo the "disable gc regions" change that, as I understand it, was to trigger testing of segments on x64 -- will revert it once it's green.

**Comment by EgorBo** [comment]
Outerloop failures are https://github.com/dotnet/runtime/issues/76511, gcstress are pretty clean except one build error/timeout.

Reverted the test change to disable gc regions.

**Review by Maoni0 — APPROVED** [review]
I've changed what I wrote in the commit msgs I pushed slightly to reflect the final state. please use this when you merge - 

```
+ we need to proactively go mark all the in range ro segs because these objects' life time isn't accurately
expressed. The expectation is all objects on ro segs are reported as marked. 

+ fixed an existing bug - for full blocking GCs it's not enough to handle ro segs when gen start seg is 
not ephemeral, it needs to be checking for condemned gen because you can still have ro segs in range 
even with one segment per heap.

+ got rid of some unnecessary checks. 

+ I do make the assumption that ro segs are always threaded at 
the beginning of gen2 so we can exit as soon as we see an rw seg. we actually always thread these into 
heap 0's gen2 seg list but I can see a chance of that changing, if we want a better balancing of work 
done for ro segs.

+ a bit of code refactoring since now we have more code related to ro segs.
```
